### PR TITLE
ENT-4595 Add git based version

### PR DESCRIPTION
**Pull request by blsemo**
This adds the ability to publish to a git based version based on the command line property `-PversionFromGit`
@lockathan has asked for this to be added to 4.3 and 4.4 to fix some issue with version links as well. This will supersede https://github.com/corda/corda/pull/5798 by a merge from 4.3 to 4.4

### CODE REVIEW

**Review by lockathan — DISMISSED** [review]
LGTM

**Review by lockathan — DISMISSED** [review]
LGTM

File: build.gradle [hunk 1]
```diff
@@ -198,6 +201,20 @@ apply plugin: 'maven-publish'
...
+// the latest git commit hash and timestamp and create a version postfix from that
+if (project.hasProperty("versionFromGit")){
+    ext.versionSuffix = "${grgit.head().dateTime.format("yyyyMMdd_HHmmss")}-${grgit.head().abbreviatedId}"
+}
+
+// Need the `toString()` call on these, because they need to be converted from GStringImpl to Java Strings.
+if (ext.versionSuffix != ""){
+    ext.corda_release_version = "${ext.baseVersion}-${ext.versionSuffix}".toString()
+} else {
+    ext.corda_release_version = "${ext.baseVersion}".toString()
```
Comment (chrisr3) [inline build.gradle:215]:
```gradle
ext.corda_release_version = ext.baseVersion
```

**Review by chrisr3 — COMMENTED** [review]
Does this break IntelliJ native builds, i.e. does it _require_ people now to delegate to Gradle?

IIRC the generated file was checked into the repository for a reason...

**Comment by mnesbit** [comment]
I need the Chris to bless this

**Comment by chrisr3** [comment]
> I need the Chris to bless this

Hmm, tricky. First fetch me the Wicked Witch of the West's broomstick!

**Comment by lockathan** [comment]
> Does this break IntelliJ native builds, i.e. does it _require_ people now to delegate to Gradle?
> 
> IIRC the generated file was checked into the repository for a reason...

Christian said that it was automatically generated during the build, but yes, I don't know what effect this would have on IntelliJ. 

He removed the generated file because it had the side effect of hiding problems. With the file in existence, he wasn't aware that something had broken until the TC builds failed. But it worked fine on his own computer. As soon as he deleted the file on his own computer, the build tried to generate it again and failed, causing him to be able to replicate the problem that TC was getting.

So the problem is that TC doesn't actually use what is checked in - it always tries to generate the file again, and if it fails, you get this weird exception saying it could not find the file. This means that Corda builds differently locally vs in TC, which is why he wanted to delete the file.

**Comment by chrisr3** [comment]
> He removed the generated file because it had the side effect of hiding problems. With the file in existence, he wasn't aware that something had broken until the TC builds failed. But it worked fine on his own computer. As soon as he deleted the file on his own computer, the build tried to generate it again and failed, causing him to be able to replicate the problem that TC was getting.

TC builds using Gradle, so if the build was working locally for Christian and failing on TC then it sounds like Christian _doesn't_ delegate to Gradle ... :worried:.

File: common/logging/src/main/template/kotlin/net/corda/common/logging/Constants.kt [hunk 2]
```diff
@@ -5,7 +5,7 @@ package net.corda.common.logging
 /**
  * constants in this file are generated by gradle
  * to change this file, edit src/main/template/kotlin/net/corda/common/logging/Constants.kt
- * the generated file does not need to be committed to source control (originally added to source control for ease of use)
+ * the generated file does not need to be committed to source control
```
Comment (chrisr3) [inline common/logging/src/main/template/kotlin/net/corda/common/logging/Constants.kt]: Can we restore this comment please? Or maybe update it to mention the IntelliJ native build?

**Review by chrisr3 — CHANGES_REQUESTED** [review]
I think we should keep the comment in the template file.

**Review by chrisr3 — APPROVED** [review]
LGTM - although @blsemo may pull more of his hair out.

**Pull request merged**